gtsam/constrained/AugmentedLagrangianOptimizer.cpp

```diff
@@ -0,0 +1,228 @@
+
+ * -------------------------------------------------------------------------- */
+
+/**
+ * @file    AugmentedLagrangianOptimizer.h
+ * @brief   Augmented Lagrangian method for nonlinear constrained optimization.
+ * @author  Yetong Zhang
+ * @date    Aug 3, 2024
+ */
+
+#include <gtsam/constrained/AugmentedLagrangianOptimizer.h>
+#include <gtsam/constrained/BiasedFactor.h>
+#include <gtsam/slam/AntiFactor.h>
+
+#include <iomanip>
+
+using std::setw, std::cout, std::endl, std::setprecision;
+
+namespace gtsam {
+
+/* ************************************************************************* */
+void AugmentedLagrangianState::initializeLagrangeMultipliers(
+    const ConstrainedOptProblem& problem) {
+  lambda_e = std::vector<Vector>();
+  lambda_e.reserve(problem.eConstraints().size());
+  for (const auto& constraint : problem.eConstraints()) {
+    lambda_e.push_back(Vector::Zero(constraint->dim()));
+  }
+  lambda_i = std::vector<double>(problem.iConstraints().size(), 0);
+}
+
+/* ************************************************************************* */
+std::tuple<AugmentedLagrangianOptimizer::State, double, double>
+AugmentedLagrangianOptimizer::iterate(const State& state, const double mu_e,
+                                      const double mu_i) const {
+  State new_state(state.iteration + 1);
+  new_state.mu_e = mu_e;
+  new_state.mu_i = mu_i;
+
+  // Update Lagrangian multipliers.
+  updateLagrangeMultiplier(state, new_state);
+
+  // Construct merit function.
+  NonlinearFactorGraph dual_graph = LagrangeDualFunction(new_state);
+
+  // Run unconstrained optimization.
+  auto optimizer = createUnconstrainedOptimizer(dual_graph, state.values);
+  new_state.setValues(optimizer->optimize(), problem_);
+  new_state.unconstrained_iters = optimizer->iterations();
+
+  // Update penalty parameters for next iteration.
+  double next_mu_e, next_mu_i;
+  std::tie(next_mu_e, next_mu_i) = updatePenaltyParameter(state, new_state);
+
+  return {new_state, next_mu_e, next_mu_i};
+}
+
+/* ************************************************************************* */
+Values AugmentedLagrangianOptimizer::optimize() const {
+  /// Construct initial state
+  State prev_state;
+  State state(0, init_values_, problem_);
+  state.initializeLagrangeMultipliers(problem_);
+  LogInit(state);
+
+  /// Set penalty parameters for the first iteration.
+  double mu_e = p_->initial_mu_e;
+  double mu_i = p_->initial_mu_i;
+
+  /// iterates
+  do {
+    prev_state = std::move(state);
+    std::tie(state, mu_e, mu_i) = iterate(prev_state, mu_e, mu_i);
+    LogIter(state);
+  } while (!checkConvergence(state, prev_state, *p_));
+
+  return state.values;
+}
+
+/* ************************************************************************* */
+NonlinearFactorGraph AugmentedLagrangianOptimizer::LagrangeDualFunction(
+    const State& state, const double epsilon) const {
+  // Initialize by adding in cost factors.
+  NonlinearFactorGraph graph = problem_.costs();
```

Comment:
If costs are *linear* and big then this could be vey inefficient, as we would be "linearizing" and then factorizing the same thing over and over. So, this might suggest that for linear const functions (with either linear or non-linear constraints) we might need to do something special.In afc, if constraints and linear as well, this is just QP. So, thinking about relationship with current QP solve is also a TODO, I guess.

Reply:
If all constraints are linear and the initial value is feasible, we can directly use the existing NonlinearOptimizer with constrained noisemodel to specify the constraints. This works because
* The existing GTSAM code can solve a constrained linear least squares problem
* If all constraints are linear, taking a step in a direction satisfying the constraints will always land on a feasible value.